### PR TITLE
Feat/visualize chassis state

### LINKED ISSUE #256: Visualize Chassis State in RVIZ
**Goal:** A [RVIZ](http://wiki.ros.org/rviz) panel to visualize [chassis_state message](https://github.com/RoboJackets/roboracing-software/blob/master/rr_msgs/msg/chassis_state.msg) data including steering, speed, and control board state to allow us to quickly determine how the car control board is doing without reading or echoing the data directly.

**Notes:** The control board reports back steering, speed, and other data, and this is published by a relay node on the topic `/chassis_state`
The speed should be visualized as a car speedometer, the steering visualized as an arrow and/or a steering wheel, and the board state as words with highlight color. The control board state is reported as a number, which represents a state such as `E_STOPPED`, `IDLE`, etc. If the state is `E_STOPPED`, this could be highlighted in red and similar color scheme for other states.
Feel free to experiment with other possibly better ways to visualize this information.

RVIZ panels use the [QT library](https://doc.qt.io/qt-5/) to draw.
There is a readme [HERE](https://github.com/RoboJackets/roboracing-software/blob/master/rr_rviz_plugins/HOWTO.md) that explains how to set up an RVIZ panel, and you can see the [ExamplePanel](https://github.com/RoboJackets/roboracing-software/blob/master/rr_rviz_plugins/src/ExamplePanel.cpp) for a quick start panel (Do not modify ExamplePanel!).

**New Files Created (at least):** `/rr_rviz_plugins/include/rr_rviz_plugins/chassisPanel.h` , `/rr_rviz_plugins/src/chassisPanel.cpp`
**Files Edited (at least):** `/rr_rviz_plugins/plugin_description.xml` , `/rr_rviz_plugins/CMakeLists.txt`

**Returning Member Mentor:** Brian Cochran @btdubs --Ask him questions!
When you believe you have finished the success criteria below, submit a [pull request](https://help.github.com/en/articles/about-pull-requests) to merge your code into master.

**Success Criteria:**
- Panel is available to open in RVIZ
- Displays notification when no `/chassis_state` published
- Speedometer speed is accurate with reported speed
- Steering visualization is accurate with  reported steering angle
- Steering viz and speedometer have numeric associated with the visualization
- State of control board is visualized with color

**Future Work (if you want to give it a try, but don't waste too much time on it):**
- Allow dynamic changing of speedometer range
- Show a small section with average speed and average steering over the past few seconds

### PR DESCRIPTION
**Pull request by NicoBartholomai**
Fixes #256  Adds a widget to RVIZ called ChassisPanel which shows the heading and speed that /chassis_state is reporting as well as weather it is recording a bag file and the estop is on or off.

### CODE REVIEW

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    painter.eraseRect(x_pos, y_pos + (height / 2), width + 2, height / 2 + 2);
+    // top speed displayed by the speedometer
+    int top_speed = 40;
+    // only changes the speed shown by the speedometer, so it does not go below its half circle,
+    // does not affect what the label says
+    if (cur_speed >= top_speed) {
+        cur_speed = top_speed;
+    }
+    // the angle the speedometer needle needs to point in
+    auto angle = (((top_speed - cur_speed) * M_PI) / top_speed);
```
[the remark this answers is not in the document]
Comment (Daniel-Martin576) [reply rr_rviz_plugins/src/ChassisPanel.cpp]: Also remember that sometimes we go backward, aka speed_mps is negative.

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+void ChassisPanel::chassisStateCallback(const rr_msgs::chassis_state msg) {
+    // for timerCallback
+    chassisUpdateTime = ros::Time::now();
+    received_signal = true;
+
+    // global variables
+    cur_speed = msg.speed_mps;
+    hding = msg.steer_rad;
+    record_bag = msg.record_bag;
+    estop = msg.estop_on;
```
Comment (Daniel-Martin576) [inline rr_rviz_plugins/src/ChassisPanel.cpp]: Personally, I think the name like steer_rad is more clear than hding. Maybe just use the same names that chassis state uses

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    // label updates
+    auto speed = spacing.toStdString() + std::to_string(msg.speed_mps) + " m/s";
```
Comment (Daniel-Martin576) [inline rr_rviz_plugins/src/ChassisPanel.cpp]: std::string speed_str might be a better variable name

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        // rcding_bag_label->setStyleSheet("QLabel { background-color : green}");
+    } else {
+        rcding_bag_label->setText("Recording Bag File:  false");
+        // rcding_bag_label->setStyleSheet("QLabel { background-color : red}");
+    }
+    if (msg.estop_on) {
+        estop_label->setText("Estop: on");
+    } else {
+        estop_label->setText("Estop: off");
+    }
```
Comment (Daniel-Martin576) [inline rr_rviz_plugins/src/ChassisPanel.cpp]: You probably can convert this to two lines just like how you did for the other ones.

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+bool estop;
+QLabel *speed_label;
+QLabel *heading_label;
+QLabel *msg_label;
+QLabel *rcding_bag_label;
+QLabel *estop_label;
+ros::Time now;
+ros::Time chassisUpdateTime = ros::Time::now();
+ros::Timer time;
+QString spacing = "          ";
```
Comment (Daniel-Martin576) [inline rr_rviz_plugins/src/ChassisPanel.cpp]: Probably should be const. Not the biggest fan of just adding spacing, maybe should try to shift the label itself by setting its position.

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    // latency for detecting if its no longer sending a message
+    if (now - chassisUpdateTime > ros::Duration(.5)) {
+        received_signal = false;
+        cur_speed = 0;
+        hding = 0;
+        speed_label->setText(spacing + "No Message");
+        heading_label->setText(spacing + "No Message");
+        msg_label->setText("No Message");
+        rcding_bag_label->setText("No Message");
+        estop_label->setText("No Message");
```
[the remark this answers is not in the document]
Comment (NicoBartholomai) [reply rr_rviz_plugins/src/ChassisPanel.cpp]: I am not getting a red and green box for my no message signal, what are you feeding it to produce that?

[the remark this answers is not in the document]
Comment (NicoBartholomai) [reply rr_rviz_plugins/src/ChassisPanel.cpp]: Okay so after I changed the Timer to be local, I also saw that issue with the colors being there unintentionally, increasing the duration for the timer seemed to fix that.

File: rr_rviz_plugins/src/ChassisPanel.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    auto *layout = new QVBoxLayout;
+
+    speed_label = new QLabel(spacing + "No Message");
+    heading_label = new QLabel(spacing + "No Message");
+    msg_label = new QLabel("No Message");
+    rcding_bag_label = new QLabel("No Message");
+    estop_label = new QLabel("No Message");
+
+    chassis_sub = nh.subscribe<rr_msgs::chassis_state>("/chassis_state", 1, ChassisPanel::chassisStateCallback);
+    time = nh.createTimer(ros::Duration(1), ChassisPanel::timerCallback);
```
[the remark this answers is not in the document]
Comment (NicoBartholomai) [reply rr_rviz_plugins/src/ChassisPanel.cpp]: When time becomes a local variable, its callback function does not seem to run. Earlier in some comments further down I though this was a duration issue, but that does not seem to be the case so I am keeping it as a global variable for now.

File: rr_rviz_plugins/src/ChassisPanel.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (NicoBartholomai) [reply rr_rviz_plugins/src/ChassisPanel.cpp]: Okay, the speedometer only shows the abs val of the speed now, similar to how a car would in reverse. The label would still say that the speed is negative.

File: rr_rviz_plugins/src/ChassisPanel.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Daniel-Martin576) [reply rr_rviz_plugins/src/ChassisPanel.cpp]: You are right, it needs to be a global variable so the timer is persistent. Made the comment while experimenting and forgot to delete it. Good catch.

**Pull request merged**